src/lib/CofEnv.ml

```diff
@@ -4,55 +4,108 @@ module D = Domain
 module UF = DisjointSet.Make (PersistentTable.M)
 module VarSet = Set.Make (Int)
 
-type env =
+type reduced_env =
   { classes : D.dim UF.t;
-    (** equivalence classes of dimensions *)
+    (** equivalence classes of dimensions from reduced cofibrations *)
 
-    cof : D.cof;
-    (** a cofibration which is assumed true, but has not been eliminated yet *)
+    true_vars : VarSet.t
+  }
+
+type env' =
+  { classes : D.dim UF.t;
+    (** equivalence classes of dimensions from reduced cofibrations *)
 
-    status : [`Consistent | `Inconsistent];
-    (** a cache which must be maintained *)
+    true_vars : VarSet.t;
 
-    true_vars : VarSet.t
+    unreduced_joins : D.cof list list;
+    (** a stack of unreduced joins, each represented by a list of cofibrations *)
   }
 
-let init () =
-  {classes = UF.init ~size:100;
-   true_vars = VarSet.empty;
-   cof = Cof.top;
-   status = `Consistent}
+type env = [ `Consistent of env' | `Inconsistent ]
+
+let init () = `Consistent
+    {classes = UF.init ~size:100;
+     true_vars = VarSet.empty;
+     unreduced_joins = []}
 
-let inconsistent =
-  {classes = UF.init ~size:0;
-   cof = Cof.bot;
-   true_vars = VarSet.empty;
-   status = `Inconsistent}
+let inconsistent = `Inconsistent
 
-let status env = env.status
+let is_consistent =
```

Comment:
What is the motivation for renaming this from `status`? I would expect any function that begins with `is_` to return a boolean.

In this case, if you dislike `status` we could call it `consistency`.

Reply:
Let's call it `consistency`. I just felt the name `status` is not very informative. We can also do `bool`, but I really like the extremely informative tags---it's impossible to mess up the branches in `abort_if_inconsistent`. I am worried about owning myself with double negation.

Reply:
sounds good!